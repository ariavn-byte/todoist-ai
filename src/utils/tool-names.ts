/**
 * Centralized tool names module
 *
 * This module provides a single source of truth for all tool names used throughout the codebase.
 * Each tool should import its own name from this module to avoid hardcoded strings.
 * This prevents outdated references when tool names change.
 */
export const ToolNames = {
    // Task management tools
    ADD_TASKS: 'add-tasks',
    COMPLETE_TASKS: 'complete-tasks',
    UPDATE_TASKS: 'update-tasks',
    FIND_TASKS: 'find-tasks',
    FIND_TASKS_BY_DATE: 'find-tasks-by-date',
    FIND_COMPLETED_TASKS: 'find-completed-tasks',

    // Project management tools
    ADD_PROJECTS: 'add-projects',
    UPDATE_PROJECTS: 'update-projects',
    FIND_PROJECTS: 'find-projects',

    // Section management tools
    ADD_SECTIONS: 'add-sections',
    UPDATE_SECTIONS: 'update-sections',
    FIND_SECTIONS: 'find-sections',

    // General tools
    GET_OVERVIEW: 'get-overview',
    DELETE_OBJECT: 'delete-object',
} as const

// Type for all tool names
export type ToolName = (typeof ToolNames)[keyof typeof ToolNames]
