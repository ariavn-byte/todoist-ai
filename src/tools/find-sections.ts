import type { Section } from '@doist/todoist-api-typescript'
import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { summarizeList } from '../utils/response-builders.js'
import { ToolNames } from '../utils/tool-names.js'

const { ADD_SECTIONS, UPDATE_SECTIONS, FIND_TASKS, UPDATE_TASKS, DELETE_OBJECT } = ToolNames

const ArgsSchema = {
    projectId: z.string().min(1).describe('The ID of the project to search sections in.'),
    search: z
        .string()
        .optional()
        .describe(
            'Search for a section by name (partial and case insensitive match). If omitted, all sections in the project are returned.',
        ),
}

type SectionSummary = {
    id: string
    name: string
}

const findSections = {
    name: ToolNames.FIND_SECTIONS,
    description: 'Search for sections by name or other criteria in a project.',
    parameters: ArgsSchema,
    async execute(args, client) {
        const { results } = await client.getSections({
            projectId: args.projectId,
        })
        const searchLower = args.search ? args.search.toLowerCase() : undefined
        const filtered = searchLower
            ? results.filter((section: Section) => section.name.toLowerCase().includes(searchLower))
            : results

        const sections = filtered.map((section) => ({
            id: section.id,
            name: section.name,
        }))

        const textContent = generateTextContent({
            sections,
            projectId: args.projectId,
            search: args.search,
        })

        return getToolOutput({
            textContent,
            structuredContent: {
                sections,
                totalCount: sections.length,
                appliedFilters: args,
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

function generateTextContent({
    sections,
    projectId,
    search,
}: {
    sections: SectionSummary[]
    projectId: string
    search?: string
}): string {
    const zeroReasonHints: string[] = []

    if (search) {
        zeroReasonHints.push('Try broader search terms')
        zeroReasonHints.push('Check spelling')
        zeroReasonHints.push('Remove search to see all sections')
    } else {
        zeroReasonHints.push('Project has no sections yet')
        zeroReasonHints.push(`Use ${ADD_SECTIONS} to create sections`)
    }

    // Data-driven next steps based on results
    const nextSteps: string[] = []

    if (sections.length > 0) {
        // Suggestions based on number of sections found
        if (sections.length === 1) {
            const sectionId = sections[0]?.id
            nextSteps.push(`Use ${FIND_TASKS} with sectionId=${sectionId} to see tasks`)
            nextSteps.push(`Use ${ADD_SECTIONS} to create additional sections for organization`)
        } else if (sections.length > 8) {
            nextSteps.push(
                'Consider consolidating sections - many small sections can reduce productivity',
            )
            nextSteps.push(`Use ${UPDATE_TASKS} to move tasks between sections`)
            nextSteps.push(`Use ${DELETE_OBJECT} with type=section to delete empty sections`)
        } else {
            nextSteps.push(`Use ${FIND_TASKS} with sectionId to see tasks in specific sections`)
            nextSteps.push(`Use ${UPDATE_SECTIONS} to modify section names`)
        }

        // Search-specific suggestions
        if (search) {
            nextSteps.push('Remove search parameter to see all sections in this project')
        }
    } else {
        // Empty result suggestions are already handled in zeroReasonHints
        // No additional nextSteps needed for empty results
    }

    const subject = search
        ? `Sections in project ${projectId} matching "${search}"`
        : `Sections in project ${projectId}`

    const previewLines =
        sections.length > 0
            ? sections.map((section) => `    ${section.name} â€¢ id=${section.id}`).join('\n')
            : undefined

    return summarizeList({
        subject,
        count: sections.length,
        previewLines,
        zeroReasonHints,
        nextSteps,
    })
}

export { findSections }
