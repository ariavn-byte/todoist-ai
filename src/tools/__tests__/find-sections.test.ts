import type { Section, TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import {
    TEST_ERRORS,
    TEST_IDS,
    createMockSection,
    extractStructuredContent,
    extractTextContent,
} from '../../utils/test-helpers.js'
import { ToolNames } from '../../utils/tool-names.js'
import { findSections } from '../find-sections.js'

// Mock the Todoist API
const mockTodoistApi = {
    getSections: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

const { FIND_SECTIONS, ADD_SECTIONS } = ToolNames

describe(`${FIND_SECTIONS} tool`, () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('listing all sections in a project', () => {
        it('should list all sections when no search parameter is provided', async () => {
            const mockSections: Section[] = [
                createMockSection({
                    id: TEST_IDS.SECTION_1,
                    projectId: TEST_IDS.PROJECT_TEST,
                    name: 'To Do',
                }),
                createMockSection({
                    id: TEST_IDS.SECTION_2,
                    projectId: TEST_IDS.PROJECT_TEST,
                    sectionOrder: 2,
                    name: 'In Progress',
                }),
                createMockSection({
                    id: 'section-789',
                    projectId: TEST_IDS.PROJECT_TEST,
                    sectionOrder: 3,
                    name: 'Done',
                }),
                createMockSection({
                    id: 'section-999',
                    projectId: TEST_IDS.PROJECT_TEST,
                    sectionOrder: 4,
                    name: 'Backlog Items',
                }),
            ]

            mockTodoistApi.getSections.mockResolvedValue({
                results: mockSections,
                nextCursor: null,
            })

            const result = await findSections.execute(
                { projectId: TEST_IDS.PROJECT_TEST },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getSections).toHaveBeenCalledWith({
                projectId: TEST_IDS.PROJECT_TEST,
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Sections in project')
            expect(textContent).toContain('To Do • id=')
            expect(textContent).toContain('In Progress • id=')
            expect(textContent).toContain('Done • id=')
            expect(textContent).toContain('Backlog Items • id=')

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent.sections).toHaveLength(4)
            expect(structuredContent.totalCount).toBe(4)
            expect(structuredContent.appliedFilters).toEqual({
                projectId: TEST_IDS.PROJECT_TEST,
                search: undefined,
            })
        })

        it('should handle project with no sections', async () => {
            mockTodoistApi.getSections.mockResolvedValue({
                results: [],
                nextCursor: null,
            })

            const result = await findSections.execute(
                { projectId: 'empty-project-id' },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getSections).toHaveBeenCalledWith({
                projectId: 'empty-project-id',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Project has no sections yet')
            expect(textContent).toContain(`Use ${ADD_SECTIONS} to create sections`)

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent.sections).toHaveLength(0)
            expect(structuredContent.totalCount).toBe(0)
            expect(structuredContent.appliedFilters).toEqual({
                projectId: 'empty-project-id',
                search: undefined,
            })
        })
    })

    describe('searching sections by name', () => {
        it('should filter sections by search term (case insensitive)', async () => {
            const mockSections: Section[] = [
                createMockSection({
                    id: TEST_IDS.SECTION_1,
                    projectId: TEST_IDS.PROJECT_TEST,
                    name: 'To Do',
                }),
                createMockSection({
                    id: TEST_IDS.SECTION_2,
                    projectId: TEST_IDS.PROJECT_TEST,
                    sectionOrder: 2,
                    name: 'In Progress',
                }),
                createMockSection({
                    id: 'section-789',
                    projectId: TEST_IDS.PROJECT_TEST,
                    sectionOrder: 3,
                    name: 'Done',
                }),
                createMockSection({
                    id: 'section-999',
                    projectId: TEST_IDS.PROJECT_TEST,
                    sectionOrder: 4,
                    name: 'Progress Review',
                }),
            ]

            mockTodoistApi.getSections.mockResolvedValue({
                results: mockSections,
                nextCursor: null,
            })

            const result = await findSections.execute(
                { projectId: TEST_IDS.PROJECT_TEST, search: 'progress' },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getSections).toHaveBeenCalledWith({
                projectId: TEST_IDS.PROJECT_TEST,
            })

            // Should return both "In Progress" and "Progress Review" (case insensitive partial match)
            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('matching "progress"')
            expect(textContent).toContain('In Progress • id=')
            expect(textContent).toContain('Progress Review • id=')
        })

        it('should handle search with no matches', async () => {
            const mockSections: Section[] = [
                createMockSection({
                    id: TEST_IDS.SECTION_1,
                    projectId: TEST_IDS.PROJECT_TEST,
                    name: 'To Do',
                }),
                createMockSection({
                    id: TEST_IDS.SECTION_2,
                    projectId: TEST_IDS.PROJECT_TEST,
                    sectionOrder: 2,
                    name: 'In Progress',
                }),
            ]

            mockTodoistApi.getSections.mockResolvedValue({
                results: mockSections,
                nextCursor: null,
            })

            const result = await findSections.execute(
                { projectId: TEST_IDS.PROJECT_TEST, search: 'nonexistent' },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Try broader search terms')
            expect(textContent).toContain('Check spelling')
            expect(textContent).toContain('Remove search to see all sections')
        })

        it('should handle case sensitive search correctly', async () => {
            const mockSections: Section[] = [
                createMockSection({
                    id: TEST_IDS.SECTION_1,
                    projectId: TEST_IDS.PROJECT_TEST,
                    name: 'Important Tasks',
                }),
                createMockSection({
                    id: TEST_IDS.SECTION_2,
                    projectId: TEST_IDS.PROJECT_TEST,
                    sectionOrder: 2,
                    name: 'Regular Work',
                }),
            ]

            mockTodoistApi.getSections.mockResolvedValue({
                results: mockSections,
                nextCursor: null,
            })

            const result = await findSections.execute(
                { projectId: TEST_IDS.PROJECT_TEST, search: 'IMPORTANT' },
                mockTodoistApi,
            )

            // Should match despite different case
            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('matching "IMPORTANT"')
            expect(textContent).toContain('Important Tasks • id=')
        })

        it('should handle partial matches correctly', async () => {
            const mockSections: Section[] = [
                createMockSection({
                    id: TEST_IDS.SECTION_1,
                    projectId: TEST_IDS.PROJECT_TEST,
                    name: 'Development Tasks',
                }),
                createMockSection({
                    id: TEST_IDS.SECTION_2,
                    projectId: TEST_IDS.PROJECT_TEST,
                    sectionOrder: 2,
                    name: 'Testing Tasks',
                }),
                createMockSection({
                    id: 'section-789',
                    projectId: TEST_IDS.PROJECT_TEST,
                    sectionOrder: 3,
                    name: 'Deployment',
                }),
            ]

            mockTodoistApi.getSections.mockResolvedValue({
                results: mockSections,
                nextCursor: null,
            })

            const result = await findSections.execute(
                { projectId: TEST_IDS.PROJECT_TEST, search: 'task' },
                mockTodoistApi,
            )

            // Should match both sections with "task" in the name
            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('matching "task"')
            expect(textContent).toContain('Development Tasks • id=')
            expect(textContent).toContain('Testing Tasks • id=')
        })

        it('should handle exact matches', async () => {
            const mockSections: Section[] = [
                createMockSection({
                    id: TEST_IDS.SECTION_1,
                    projectId: TEST_IDS.PROJECT_TEST,
                    name: 'Done',
                }),
                createMockSection({
                    id: TEST_IDS.SECTION_2,
                    projectId: TEST_IDS.PROJECT_TEST,
                    sectionOrder: 2,
                    name: 'Done Soon',
                }),
            ]

            mockTodoistApi.getSections.mockResolvedValue({
                results: mockSections,
                nextCursor: null,
            })

            const result = await findSections.execute(
                { projectId: TEST_IDS.PROJECT_TEST, search: 'done' },
                mockTodoistApi,
            )

            // Should match both sections containing "done"
            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('matching "done"')
            expect(textContent).toContain('Done • id=')
            expect(textContent).toContain('Done Soon • id=')
        })
    })

    describe('error handling', () => {
        it.each([
            { error: 'API Error: Project not found', projectId: 'non-existent-project' },
            { error: TEST_ERRORS.API_UNAUTHORIZED, projectId: 'restricted-project' },
            { error: 'API Error: Invalid project ID format', projectId: 'invalid-id-format' },
        ])('should propagate $error', async ({ error, projectId }) => {
            mockTodoistApi.getSections.mockRejectedValue(new Error(error))
            await expect(findSections.execute({ projectId }, mockTodoistApi)).rejects.toThrow(error)
        })
    })
})
