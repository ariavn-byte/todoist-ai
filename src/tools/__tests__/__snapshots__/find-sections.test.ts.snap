// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`find-sections tool listing all sections in a project should handle project with no sections 1`] = `
"Sections in project empty-project-id: 0.
No results. Project has no sections yet; Use add-sections to create sections."
`;

exports[`find-sections tool listing all sections in a project should list all sections when no search parameter is provided 1`] = `
"Sections in project 6cfCcrrCFg2xP94Q: 4.
Preview:
    To Do • id=section-123
    In Progress • id=section-456
    Done • id=section-789
    Backlog Items • id=section-999
Next:
- Use find-tasks with sectionId to see tasks in specific sections
- Use update-sections to modify section names"
`;

exports[`find-sections tool searching sections by name should filter sections by search term (case insensitive) 1`] = `
"Sections in project 6cfCcrrCFg2xP94Q matching "progress": 2.
Preview:
    In Progress • id=section-456
    Progress Review • id=section-999
Next:
- Use find-tasks with sectionId to see tasks in specific sections
- Use update-sections to modify section names
- Remove search parameter to see all sections in this project"
`;

exports[`find-sections tool searching sections by name should handle case sensitive search correctly 1`] = `
"Sections in project 6cfCcrrCFg2xP94Q matching "IMPORTANT": 1.
Preview:
    Important Tasks • id=section-123
Next:
- Use find-tasks with sectionId=section-123 to see tasks
- Use add-sections to create additional sections for organization
- Remove search parameter to see all sections in this project"
`;

exports[`find-sections tool searching sections by name should handle exact matches 1`] = `
"Sections in project 6cfCcrrCFg2xP94Q matching "done": 2.
Preview:
    Done • id=section-123
    Done Soon • id=section-456
Next:
- Use find-tasks with sectionId to see tasks in specific sections
- Use update-sections to modify section names
- Remove search parameter to see all sections in this project"
`;

exports[`find-sections tool searching sections by name should handle partial matches correctly 1`] = `
"Sections in project 6cfCcrrCFg2xP94Q matching "task": 2.
Preview:
    Development Tasks • id=section-123
    Testing Tasks • id=section-456
Next:
- Use find-tasks with sectionId to see tasks in specific sections
- Use update-sections to modify section names
- Remove search parameter to see all sections in this project"
`;

exports[`find-sections tool searching sections by name should handle search with no matches 1`] = `
"Sections in project 6cfCcrrCFg2xP94Q matching "nonexistent": 0.
No results. Try broader search terms; Check spelling; Remove search to see all sections."
`;
